File: test/payload_generator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sslh_sshl/services/payload_generator.dart';

void main() {
  group('PayloadGenerator Tests', () {
    test('parsePayload replaces tags correctly', () {
      const template =
          'GET / HTTP/1.1[crlf]Host: [host][crlf]Status: [status][crlf][crlf]';
      final parsed = PayloadGenerator.parsePayload(
        template,
        host: 'id1.jagoanip.my.id',
        port: 443,
        statusLine: 'HTTP/1.1 200 Connection established',
      );

      expect(
        parsed,
        'GET / HTTP/1.1\r\nHost: id1.jagoanip.my.id\r\nStatus: HTTP/1.1 200 Connection established\r\n\r\n',
      );
    });

    test('generatePayload with Normal injection', () {
      final options = PayloadGeneratorOptions(
        url: 'id1.jagoanip.my.id',
        method: 'GET',
        injectionMethod: 'Normal',
        upgradeWebsocket: true,
      );

      final payload = PayloadGenerator.generatePayload(options);
      expect(
        payload,
        'GET / HTTP/1.1[crlf]Host: id1.jagoanip.my.id[crlf]Upgrade: websocket[crlf][crlf]',
      );
    });

    test('generatePayload with Front Inject', () {
      final options = PayloadGeneratorOptions(
        url: 'id1.jagoanip.my.id',
        method: 'GET',
        injectionMethod: 'Front Inject',
        upgradeWebsocket: true,
      );

      final payload = PayloadGenerator.generatePayload(options);
      expect(
        payload,
        'GET http://id1.jagoanip.my.id/ HTTP/1.1[crlf]Host: id1.jagoanip.my.id[crlf]Upgrade: websocket[crlf][crlf]',
      );
    });
  });
}

File: app/src/test/kotlin/com/sslh/sshl/app/PayloadGeneratorTest.kt
package com.sslh.sshl.app

import com.sslh.sshl.app.service.PayloadGenerator
import com.sslh.sshl.app.service.PayloadGeneratorOptions
import org.junit.Assert.*
import org.junit.Test

class PayloadGeneratorTest {

    @Test
    fun testParsePayloadTags() {
        val rawPayload = "GET / HTTP/1.1[crlf]Host: [host][crlf]Upgrade: websocket[crlf][crlf]"
        val parsed = PayloadGenerator.parsePayload(rawPayload, host = "example.com", port = 443)

        assertTrue(parsed.contains("\r\n"))
        assertTrue(parsed.contains("Host: example.com"))
        assertFalse(parsed.contains("[crlf]"))
        assertFalse(parsed.contains("[host]"))
    }

    @Test
    fun testGeneratePayloadNormalMode() {
        val options = PayloadGeneratorOptions(
            url = "example.com",
            method = "GET",
            injectionMethod = "Normal",
            upgradeWebsocket = true,
            keepAlive = true,
            userAgent = true
        )
        val generated = PayloadGenerator.generatePayload(options)

        assertTrue(generated.contains("GET / HTTP/1.1[crlf]"))
        assertTrue(generated.contains("Host: example.com[crlf]"))
        assertTrue(generated.contains("Upgrade: websocket[crlf]"))
        assertTrue(generated.contains("Connection: Keep-Alive[crlf]"))
        assertTrue(generated.contains("User-Agent: [ua][crlf]"))
    }

    @Test
    fun testGeneratePayloadFrontInjectMode() {
        val options = PayloadGeneratorOptions(
            url = "example.com",
            method = "POST",
            injectionMethod = "Front Inject"
        )
        val generated = PayloadGenerator.generatePayload(options)

        assertTrue(generated.contains("POST http://example.com/ HTTP/1.1[crlf]"))
    }

    @Test
    fun testGeneratePayloadBackInjectMode() {
        val options = PayloadGeneratorOptions(
            url = "example.com",
            method = "GET",
            injectionMethod = "Back Inject"
        )
        val generated = PayloadGenerator.generatePayload(options)

        assertTrue(generated.contains("CONNECT [host_port] HTTP/1.1[crlf]"))
        assertTrue(generated.contains("GET http://example.com/ HTTP/1.1[crlf]"))
    }
}
